File: tests/examples/template.cpp
#include <stdint.h>

#include <map>
#include <string>
#include <vector>

class Foo {
public:
    int a_int;
};

class UseStd {
    std::map<uint8_t, std::pair<Foo, bool>> a_map;
};

template <class T>
class Template {
    T t;
};

class UseTemplate {
    Template<uint8_t> templ;
};

template <typename Type, typename SubType = int>
class TypenameTemplate {
    std::map<Type, SubType> t;
};

class UseTypenameTemplate {
    TypenameTemplate<uint8_t, bool> templ;
};

namespace Value {
template <template <typename U, typename... Args> class ArrayType = std::vector,
          class StringType                                        = std::string>
class ValueBase {
    using array_t  = ArrayType<ValueBase>;
    using string_t = StringType;
    union value {
        array_t *a;
        string_t *s;
    };
    value v;
};
using ValueClass = ValueBase<>;

template <typename T>
class MyVector : public std::vector<T> {
    MyVector() : std::vector<T>() {}
};

class UseValueClass {
    ValueBase<MyVector, Foo> sc;
};

};  // namespace Value

File: tests/examples/template.py
from examples.utils import *

def case_template():
    data = read_json('template.cpp.json')
    expect_eq(data['UseTemplate']['fields'][0]['templates'][0]['type'] , 'Template')
    expect_eq(data['UseTemplate']['fields'][0]['types'][0]['type'] , 'unsigned char')
    expect_true(any((t['type'] == 'map' for t in data['UseStd']['fields'][0]['templates'])))
    expect_true(any((t['type'] == 'pair' for t in data['UseStd']['fields'][0]['templates'])))
    expect_true(any((t['type'] == 'unsigned char' for t in data['UseStd']['fields'][0]['types'])))
    expect_true(any((t['type'] == 'Foo' for t in data['UseStd']['fields'][0]['types'])))
    expect_true(any((t['type'] == '_Bool' for t in data['UseStd']['fields'][0]['types'])))
    expect_false(data['Template']['fields'][0]['isReduced'])
    expect_true(data['Template']['fields'][0]['types'][0]['isReduced'])
    expect_eq(data['Template']['templates'][0] , 'T')
    expect_true(any((t == 'Type' for t in data['TypenameTemplate']['templates'])))
    expect_true(any((t == 'SubType' for t in data['TypenameTemplate']['templates'])))
